helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java

```diff
@@ -93,42 +184,126 @@ public static ClusterModel generateClusterModel(ResourceControllerDataProvider d
+          getValidStateInstanceMap(baselineAssignment.get(resourceName), activeInstances);
+      Map<String, Map<String, Set<String>>> bestPossiblePartitionStateMap =
+          getValidStateInstanceMap(bestPossibleAssignment.get(resourceName), activeInstances);
+      // Iterate the replicas of the resource to find the ones that require reallocating.
+      for (AssignableReplica replica : replicaMap.get(resourceName)) {
+        String partitionName = replica.getPartitionName();
+        String replicaState = replica.getReplicaState();
+        Set<String> baselineAllocations =
+            baselinePartitionStateMap.getOrDefault(partitionName, Collections.emptyMap())
+                .getOrDefault(replicaState, Collections.emptySet());
+        Set<String> bestPossibleAllocations =
+            bestPossiblePartitionStateMap.getOrDefault(partitionName, Collections.emptyMap())
+                .getOrDefault(replicaState, Collections.emptySet());
+
+        // Compare the best possible assignments with the baseline assignment for the common part.
+        List<String> commonAllocations = new ArrayList<>(bestPossibleAllocations);
+        commonAllocations.retainAll(baselineAllocations);
+        if (!commonAllocations.isEmpty()) {
+          // 1. If the partition is allocated at the same location in both baseline and best possible
+          // assignment, there is no need to reassign it.
+          String allocatedInstance = commonAllocations.get(0);
+          allocatedReplicas.computeIfAbsent(allocatedInstance, key -> new HashSet<>()).add(replica);
+          // Remove the instance from the record to prevent the same location being processed again.
+          baselineAllocations.remove(allocatedInstance);
+          bestPossibleAllocations.remove(allocatedInstance);
+        } else if (!baselineAllocations.isEmpty()) {
+          // 2. If the partition is allocated at an active instance in the Baseline but the
+          // allocation does not exist in the best possible assignment, try to rebalance it.
+          toBeAssignedReplicas.add(replica);
+          // Remove the instance from the baseline record to prevent the same location being picked
+          // up again.
+          baselineAllocations.remove(baselineAllocations.iterator().next());
+        } else if (!bestPossibleAllocations.isEmpty()) {
+          // 3. If the partition is allocated at an active instance in the best possible assignment
+          // only, there is no need to rebalance it.
+          String allocatedInstance = bestPossibleAllocations.iterator().next();
+          allocatedReplicas.computeIfAbsent(allocatedInstance, key -> new HashSet<>()).add(replica);
+          // Remove the instance from the record to prevent the same location being processed again.
+          bestPossibleAllocations.remove(allocatedInstance);
+        } else {
+          // 4. If the partition is completely new, rebalance it.
+          toBeAssignedReplicas.add(replica);
```

Comment:
Let's also add what's going on conceptually here:
---------------------
This replica is not found in either the baseline or the bestPossibleAssignment. This implies that the replica was newly added but was never assigned in reality (not in the bestPossibleAssignment) or was added so recently that it hasn't shown up in the baseline (because global baseline calculation takes longer and is asynchronous). In that case, we add it to `toBeAssignedReplicas` so that it will get assigned as a result of `partialRebalance`.

Reply:
Changed.